chain-impl-mockchain/src/ledger/ledger.rs

```diff
@@ -1709,6 +1723,9 @@ impl ApplyBlockLedger {
             }
         };
 
+        #[cfg(feature = "evm")]
+        new_ledger.evm.update_environment();
```

Comment:
nit, I do not think that invoke of the `update_environment();` in the `finish()` function, we need to update evm env before each transaction execution. So we need to move into the `apply_block()` ledger function, before applying all the fragments.

Also I think that we should modify `update_environment()` function to the following form:
```
fn update_environment(block_date: BlockDate, block_number: ChainLength, block_hash: Hash) {

...
} 
```
Most of these data we can get from the `metadata: &HeaderContentEvalContext` argument from the `apply_block()` function
```
pub fn apply_block(
        &self,
        ledger_params: LedgerParameters,
        contents: &Contents,
        metadata: &HeaderContentEvalContext,)
```
The last missing part `block_hash` which we can not get inside `apply_block` function currently I suggest to modify `HeaderContentEvalContext` struct to the following form:
```
pub struct HeaderContentEvalContext {
   #[cfg(feature = "evm")]
    pub(crate) content_hash: BlockContentHash,
    pub(crate) block_date: BlockDate,
    pub(crate) chain_length: ChainLength,
    pub(crate) content_hash: BlockContentHash,
    pub(crate) consensus_eval_context: ConsensusEvalContext,
}
```
And fill this field before invoking `apply_block` function to the corresponds block hash value

Reply:
> nit, I do not think that invoke of the `update_environment();` in the `finish()` function, we need to update evm env before each transaction execution. So we need to move into the `apply_block()` ledger function, before applying all the fragments.
> 
> Also I think that we should modify `update_environment()` function to the following form:
> 
> ```
> fn update_environment(block_date: BlockDate, block_number: ChainLength, block_hash: Hash) {
> 
> ...
> } 
> ```
> 
> Most of these data we can get from the `metadata: &HeaderContentEvalContext` argument from the `apply_block()` function
> 
> ```
> pub fn apply_block(
>         &self,
>         ledger_params: LedgerParameters,
>         contents: &Contents,
>         metadata: &HeaderContentEvalContext,)
> ```
> 
> The last missing part `block_hash` which we can not get inside `apply_block` function currently I suggest to modify `HeaderContentEvalContext` struct to the following form:
> 
> ```
> pub struct HeaderContentEvalContext {
>    #[cfg(feature = "evm")]
>     pub(crate) content_hash: BlockContentHash,
>     pub(crate) block_date: BlockDate,
>     pub(crate) chain_length: ChainLength,
>     pub(crate) content_hash: BlockContentHash,
>     pub(crate) consensus_eval_context: ConsensusEvalContext,
> }
> ```
> 
> And fill this field before invoking `apply_block` function to the corresponds block hash value

Definitely agree with you. This should go in a separate PR.